synapse/lib/remcycle.py

```diff
@@ -0,0 +1,846 @@
+
+    :param core: Cortex used to store ingest data.  By default, a ram cortex
+                 is used.
+    :param opts: Opts applied to the object via the Config interface.
+    :param defs: Default options applied to the object via the Config
+                 interface.  Generally this would not be overridden.
+    '''
+
+    def __init__(self,
+                 core=None,
+                 opts=None,
+                 defs=HYPNOS_BASE_DEFS,
+                 *args,
+                 **kwargs):
+        s_config.Config.__init__(self,
+                                 opts=opts,
+                                 defs=defs)
+
+        self.required_keys = (NAMESPACE, DOC, APIS)
+
+        self.apis = {}
+        self.namespaces = set([])
+        self.docs = {}
+        self.global_request_headers = {}  # Global request headers per namespace
+
+        # Check configable options before we spin up any resources
+        pool_min = self.getConfOpt(MIN_WORKER_THREADS)
+        pool_max = self.getConfOpt(MAX_WORKER_THREADS)
+        if pool_min < 1 or pool_max < pool_min:
+            raise ValueError('Bad pool configuration provided.')
+
+        # Tornado Async
+        if 'ioloop' in kwargs:
+            loop = kwargs.get('ioloop')
+        else:
+            loop = t_ioloop.IOLoop()
+        self.loop = loop
+        self.client = t_http.AsyncHTTPClient(io_loop=self.loop)
+        self.iothr = self._runIoLoop()
+
+        # Synapse Async and thread pool
+        self.boss = s_async.Boss()
+        self.pool = s_threads.Pool(pool_min, maxsize=pool_max)
+
+        # Synapse Core and ingest tracking
+        if core:
+            self.core = core
+            self.core_provided = True
+        else:
+            self.core = s_cortex.openurl('ram://')
+            self.core_provided = False
+        self._api_ingests = collections.defaultdict(list)
+
+        # Setup Fini handlers
+        self.onfini(self._onHypoFini)
+
+    def __repr__(self):
+        d = {'name': self.__class__.__name__,
+             'loc': hex(id(self)),
+             'ns': list(self.namespaces),
+             'core': self.core,
+             }
+        s = '<{name} at {loc}, namespaces: {ns}, core: {core}>'.format(**d)
+        return s
+
+    @s_threads.firethread
+    def _runIoLoop(self):
+        self.loop.start()
+
+    def _onHypoFini(self):
+        # Stop the IOLoop async thread
+        self.loop.stop()
+        self.iothr.join()
+        # Stop the boss making jobs
+        self.boss.fini()
+        # Stop the consuming pool
+        self.pool.fini()
+        # Stop the cortex if we created the cortex ourselves.
+        if not self.core_provided:
+            self.core.fini()
+
+    def addConfig(self, config, reload_config=False):
+        '''
+        Register a configuration into a Hypnos object.
+
+        The Hypnos object can accept a configuration object shaped like the following:
+
+        ::
+
+            {
+              "apis": {
```

Comment:
API definitions and their ingest definitions have been converted into sequences of two-value pairs.